File: src/main/java/algorithm/lc/ReorderList.java
package algorithm.lc;

/**
 * Given a singly linked list L: L0→L1→…→Ln-1→Ln, reorder it to:
 * L0→Ln→L1→Ln-1→L2→Ln-2→…
 * 
 * You must do this in-place without altering the nodes' values.
 * 
 * For example, Given {1,2,3,4}, reorder it to {1,4,2,3}.
 * 
 */
public class ReorderList {

	class ListNode {
		int val;
		ListNode next;

		ListNode(int x) {
			val = x;
			next = null;
		}
	}

	/**
	 * O(n) time, O(1) space. 
	 *
	 */
	public class Solution {
		public void reorderList(ListNode head) {
			if (head == null || head.next == null) {
				return;
			}
			// split list into half
			ListNode slow = head;
			ListNode fast = head;
			while (fast.next != null && fast.next.next != null) {
				slow = slow.next;
				fast = fast.next.next;
			}

			ListNode firstHalf = head;
			ListNode cur = slow.next;
			slow.next = null;

			// reverse second half
			ListNode prev = null;
			while (cur != null) {
				ListNode next = cur.next;
				cur.next = prev;
				prev = cur;
				cur = next;
			}
			ListNode secondHalf = prev;

			// merge two list      
			ListNode newHead = new ListNode(0);
			cur = newHead;
			boolean addFirst = true;
			while (firstHalf != null || secondHalf != null) {
				if (addFirst) {
					cur.next = firstHalf;
					firstHalf = firstHalf.next;
				} else {
					cur.next = secondHalf;
					secondHalf = secondHalf.next;
				}
				addFirst = !addFirst;
				cur = cur.next;
			}
			head = newHead.next;
		}
	}

}

File: src/main/python/lc/reorder_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
  # @param head, a ListNode
  # @return nothing
  def reorderList(self, head):
    if head is None or head.next is None:
      return
    
    # decompose list to two halves
    slow = fast = head
    while fast.next is not None and fast.next.next is not None:
      slow = slow.next
      fast = fast.next.next
    
    first_half = head
    second_half = slow.next      
    slow.next = None
    
    # reverse second half
    prev = None
    cur = second_half
    while cur is not None:
      next = cur.next
      cur.next = prev
      prev = cur
      cur = next
    second_half = prev
    
    # merge
    new_head = ListNode(0)
    cur = new_head
    add_first = True
    while first_half is not None or second_half is not None:
      if add_first:
        cur.next = first_half
        first_half = first_half.next
      else:
        cur.next = second_half
        second_half = second_half.next
        
      add_first = not add_first
      cur = cur.next
    
    head = new_head.next
